docs: dfx extension cli & design doc (#3433)
dfinity/sdk: docs/design/dfx-extensions.md

# DFX extensions - how it works 

`dfx` extension is a feature in the DFINITY Developer Software Development Kit (SDK) that allows for extending the `dfx` CLI's core functionality. Modeled after the principles of [Git custom commands](https://mfranc.com/tools/git-custom-command/) and [Cargo's custom subcommands](https://doc.rust-lang.org/book/ch14-05-extending-cargo.html#extending-cargo-with-custom-commands), the feature enables the addition of user-defined commands that seamlessly integrate with the existing `dfx` command set. 

> ## dfx exension install

Reviewer: ```suggestion
## dfx extension install
```

    - If the extension name is not associated with the dfx version, another error is raised.
    - The versions of the extension that are compatible with the dfx version are then sorted, and the latest version is chosen as the most suitable one.
    - This mechanism ensures that users always get the newest version of an extension that works correctly with their dfx version.
    - In essence, while dfx provides tools to manage extensions, the compatibility matrix and extension manifest add an extra layer of sophistication. They ensure that extensions are accurately described, can be transformed into command-line interfaces, and most importantly, are always in sync with the version of dfx the user is operating. This comprehensive management system guarantees smooth operations and enhances user experience.

2. Download and Extraction:
    - Once the compatible version of the extension is determined, dfx constructs a download URL. This URL points to a GitHub releases page where the extensions are hosted. Currently, extensions support downloading only from `dfinity/dfx-extensions` repository. Here is how the URL template looks like:

>     ```

Reviewer: The three backticks should not be indented.  This seems to mess up the formatting going forward.

Author: I wanted the codeblock to indented together with bullet point, I made the adjustment, what do you think https://github.com/dfinity/sdk/blob/caefa3588ac04446a0c818a9c431d7da2a129673/docs/design/dfx-extensions.md

    "https://github.com/dfinity/dfx-extensions/releases/download/{github_release_tag}/{extension_archive_name}.tar.gz";
    ```

>     - The extension is then downloaded from this URL. If the download fails for any reason, an error is returned.

Reviewer: Should be no indentation for the triple-backquote
Should be a blank line after

Author: github diff-view seems to incorrectly render it
see rendered view, looks alright there https://github.com/dfinity/sdk/blob/caefa3588ac04446a0c818a9c431d7da2a129673/docs/design/dfx-extensions.md

    - After successful download, the extension, which is in a compressed archive format (.tar.gz), is unpacked to a temporary directory.

3. Final Installation:

>     - Post extraction, the extension is renamed and moved to its permanent location (the directory where extensions are meant to reside).

Reviewer: ```suggestion
    - Post-extraction, the extension is renamed and moved to its permanent location (the directory where extensions are meant to reside).
```

    - If the user specifies a custom name for the extension at the time of installation, the binary is renamed to reflect this custom name.
    - On UNIX systems, appropriate permissions are set for the installed extension to ensure it's executable.

4. Error Handling:
    - Throughout this process, any potential errors are diligently checked for, ensuring a robust installation process. Examples include errors related to incompatible versions, download failures, decompression errors, etc.
    - This process ensures that the extensions are not only installed from a trusted source but are also compatible with the user's dfx version, enhancing the utility's reliability and user experience.

## Running Extensions

Installed extensions are stored in `dfx`'s cache. Let's say you have `sns` extension installed; here is how the cache directory structure will look like:
```console
❯ tree $(dfx cache show)
~/.cache/dfinity/versions/0.15.0
└── extensions
   └── sns
      ├── CHANGELOG.md
      ├── extension.json
      ├── LICENSE
      ├── README.md
      ├── sns
      └── sns-cli
```

> When requested to run an extension, dfx fetches the location of the executable associated with the extension, by determining the cache location associated with its current version, and searching for directory with extension name inside: `$(dfx cache show)/extensions/EXT_NAME/EXT_NAME`. Fox example , when you're running `dfx sns ARGS` or `dfx extension run sns ARGS`, `dfx` will try to find and extension under this path: `$(dfx cache show)/extensions/sns/sns`. 

Reviewer: ```suggestion
When requested to run an extension, dfx fetches the location of the executable associated with the extension, by determining the cache location associated with its current version, and searching for directory with extension name inside: `$(dfx cache show)/extensions/EXT_NAME/EXT_NAME`. Fox example , when you're running `dfx sns ARGS` or `dfx extension run sns ARGS`, `dfx` will try to find the extension binary at this path: `$(dfx cache show)/extensions/sns/sns`. 
```

When executing the binary, the path to the cache is appended as an argument when invoking the extension's executable, in practice that means the extension will always be executed with `--dfx-cache-path $(dfx cache show)` parameter, like so: `exec $(dfx cache show)/extensions/sns/sns --dfx-cache-path $(dfx cache show)`. This allows the extension to depend on other extensions, and binaries stored in dfx's cache.

The extension's binary is launched as a child process. The system waits for the child process to finish executing. Once the process completes, dfx checks the exit code. If it's non-zero (indicating an error), an error is returned.

### Extensions as Commands

> Extensions are converted to command-line commands via an `extension.json` manifest, which encodes metadata and command data. Once installed, extensions produce CLI commands with defined subcommands and arguments through this manifest. The manifest file contains metadata like the extension's name, version, authors, dependencies, and subcommands. Each extension can possess subcommands with specific arguments. By translating the manifest into command-line commands, dfx enables the user to interact with extension commands, including their subcommands and arguments, which in practice means that any command and subcommand will be displayed when `dfx EXT_NAME --help`, or `dfx EXT_NAME SUBCMD` is called, and it will display all available args, options and subcommands. 

Reviewer: ```suggestion
Extensions are converted to command-line commands via an `extension.json` manifest, which encodes metadata and command data. Once installed, extensions produce CLI commands with defined subcommands and arguments through this manifest. The manifest file contains metadata like the extension's name, version, authors, dependencies, and subcommands. Each extension can provide subcommands with specific arguments. By translating the manifest into command-line commands, dfx enables the user to interact with extension commands, including their subcommands and arguments, which in practice means that any command and subcommand will be displayed when `dfx EXT_NAME --help`, or `dfx EXT_NAME SUBCMD` is called, and it will display all available args, options and subcommands. 
```

# DFX extensions - how it works 

`dfx` extension is a feature in the DFINITY Developer Software Development Kit (SDK) that allows for extending the `dfx` CLI's core functionality. Modeled after the principles of [Git custom commands](https://mfranc.com/tools/git-custom-command/) and [Cargo's custom subcommands](https://doc.rust-lang.org/book/ch14-05-extending-cargo.html#extending-cargo-with-custom-commands), the feature enables the addition of user-defined commands that seamlessly integrate with the existing `dfx` command set. 

> 

Reviewer: Maybe start by thinking about what are the things that you think a reader should understand after reading this document, and organize the document around that.

## dfx extension install

The dfx utility offers a feature to install new extensions. Here's a high-level overview of how the installation process works:

1. Determine Extension Compatibility:
    - Before any installation can take place, dfx checks if the requested extension is already installed. If it is, the process terminates with an error.
    - The utility identifies the version of the extension compatible with its own version. This ensures that users don't end up installing extensions that may not work properly with their specific version of dfx.

>     - Compatibility is determined using an extension compatibility matrix, which presumably maps extensions to compatible versions of dfx.

Reviewer: Overall, there's either too much detail here, or it's not organized in a way that it can be consumed.  I would expect a reader's response to seeing this document  be "TLDR".  As a starting point, these 13 bullet points could probably be summarized in 3 or 4 higher-level points (and overall, all shorter than this).

    - This mechanism ensures that users always get the newest version of an extension that works correctly with their dfx version.
    - In essence, while dfx provides tools to manage extensions, the compatibility matrix and extension manifest add an extra layer of sophistication. They ensure that extensions are accurately described, can be transformed into command-line interfaces, and most importantly, are always in sync with the version of dfx the user is operating. This comprehensive management system guarantees smooth operations and enhances user experience.

2. Download and Extraction:
    - Once the compatible version of the extension is determined, dfx constructs a download URL. This URL points to a GitHub releases page where the extensions are hosted. Currently, extensions support downloading only from `dfinity/dfx-extensions` repository. Here is how the URL template looks like:
        ```
        "https://github.com/dfinity/dfx-extensions/releases/download/{github_release_tag}/{extension_archive_name}.tar.gz";
        ```
        

>     - The extension is then downloaded from this URL. If the download fails for any reason, an error is returned.
>     - After successful download, the extension, which is in a compressed archive format (.tar.gz), is unpacked to a temporary directory.

Reviewer: Do these two bullet points provide useful information?
```suggestion
```

3. Final Installation:

>     - Post-extraction, the extension is renamed and moved to its permanent location (the directory where extensions are meant to reside).
>     - If the user specifies a custom name for the extension at the time of installation, the binary is renamed to reflect this custom name.
>     - On UNIX systems, appropriate permissions are set for the installed extension to ensure it's executable.

Reviewer: There's so much noise here that the important pieces will be easy to miss.
```suggestion
    - If the user specifies a custom name for the extension at the time of installation, the binary is renamed to reflect this custom name.
```

4. Error Handling:

>     - Throughout this process, any potential errors are diligently checked for, ensuring a robust installation process. Examples include errors related to incompatible versions, download failures, decompression errors, etc.

Reviewer: More back-patting: `diligently` and `robust`.  Checking for errors like download failures and decompression errors should be a given and do not need to be mentioned in this document.  This reads like filler, maybe written by chatgpt.

    - This process ensures that the extensions are not only installed from a trusted source but are also compatible with the user's dfx version, enhancing the utility's reliability and user experience.

## Running Extensions

Installed extensions are stored in `dfx`'s cache. Let's say you have `sns` extension installed; here is how the cache directory structure will look like:
```console
❯ tree $(dfx cache show)
~/.cache/dfinity/versions/0.15.0
└── extensions
   └── sns
      ├── CHANGELOG.md
      ├── extension.json
      ├── LICENSE
      ├── README.md
      ├── sns
      └── sns-cli
```

When requested to run an extension, dfx fetches the location of the executable associated with the extension, by determining the cache location associated with its current version, and searching for directory with extension name inside: `$(dfx cache show)/extensions/EXT_NAME/EXT_NAME`. Fox example , when you're running `dfx sns ARGS` or `dfx extension run sns ARGS`, `dfx` will try to find the extension binary at this path: `$(dfx cache show)/extensions/sns/sns`. 

> When executing the binary, the path to the cache is appended as an argument when invoking the extension's executable, in practice that means the extension will always be executed with `--dfx-cache-path $(dfx cache show)` parameter, like so: `exec $(dfx cache show)/extensions/sns/sns --dfx-cache-path $(dfx cache show)`. This allows the extension to depend on other extensions, and binaries stored in dfx's cache.

Reviewer: Rather than prose, it would be better to describe which parameters dfx passes to the extension binary in the form of a table or list.  A prose description / example can follow.

Otherwise, buried in the middle of the paragraph is the fact that dfx passes `--dfx-cache-path "$(dfx cache show)"`.
